### PR TITLE
feat(kit): detect pdf is not visible in browser

### LINKED ISSUE #1866: 🚀 - `PdfViewer` display message when inline PDF is disabled
### What is the affected URL?

https://taiga-ui.dev/components/pdf-viewer

### Description

Just to press any button that must show PDF.
Expected: PdfViewver with pdf in it.
Actual: pdf-file download starts.
Firefox 101, no addons. Windows 11 21H2. Google Chrome - no issue.
[Screen Record](https://disk.yandex.ru/i/OWXRlshMG5SL2g)

### Which browsers have you used?

- [ ] Chrome
- [X] Firefox
- [ ] Safari
- [ ] Edge

### Which operating systems have you used?

- [ ] macOS
- [X] Windows
- [ ] Linux
- [ ] iOS
- [ ] Android

### PR DESCRIPTION
**Pull request by splincode**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows [Conventional Commits](https://www.conventionalcommits.org/en/)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring
- [ ] Code style update
- [ ] Build or CI related changes
- [ ] Documentation content changes

## What is the current behavior?

Closes #1866

## What is the new behavior?

https://user-images.githubusercontent.com/12021443/172903333-3dc1ea34-f114-47cd-829d-06a227e70970.mov

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: projects/kit/components/pdf-viewer/pdf-viewer.template.html [hunk 1]
```diff
@@ -30,3 +30,18 @@ <h2 class="t-title">{{context.label}}</h2>
...
+<div
+    *ngIf="emptyPreview"
+    class="t-error-message"
+    @tuiFadeIn
+>
+    <ng-container *ngIf="context.errorMessage; else defaultMessage">
+        {{context.errorMessage}}
+    </ng-container>
+    <ng-template #defaultMessage>
+        The PDF could not be loaded, there may be a network problem, or your
```
Comment (nsbarsukov) [inline projects/kit/components/pdf-viewer/pdf-viewer.template.html:43]: Let's put it inside `@taiga-ui/i18n`:
https://github.com/Tinkoff/taiga-ui/blob/7b303d12621a7270e6a2b9d00651b76739486274/projects/i18n/interfaces/language.ts#L23

For example, property `unavailablePdfError` (or something shorter)

File: projects/kit/components/pdf-viewer/pdf-viewer.component.ts [hunk 2]
```diff
@@ -1,33 +1,103 @@
...
+                .pipe(
+                    map(() => this.hasPdfInsideIframe(iframe)),
+                    filter(pdfVisible => !pdfVisible),
+                    takeUntil(this.destroy$),
+                )
+                .subscribe(() => {
+                    // hide default preview (chrome/safari) for download pdf
+                    iframe.style.setProperty('display', 'none');
+                    this.emptyPreview = true;
+                    this.cd.detectChanges();
```
Comment (nsbarsukov) [inline projects/kit/components/pdf-viewer/pdf-viewer.component.ts:71]: We can deal with it without extra explicit subscription and extra selector.
Use `*tuiLet` inside template to subscribe to this observable.
Then in template:
```html
<header class="t-header" *tuiLet="iframe$ | async as iframeAvailable">
    ...
    <section
        polymorpheus-outlet
        class="t-content"
        [content]="iframeAvailable ? context.content : errorMessageRef"
        <iframe
            class="t-iframe"
            [class.t-iframe_hidden]="!iframeAvailable"
        ></iframe>
    </section>
    
    <ng-template #errorMessageRef>...</ng-template>
</header>

```

File: projects/kit/components/pdf-viewer/pdf-viewer.component.ts [hunk 3]
```diff
@@ -1,33 +1,103 @@
...
+                });
+        }
+    }
+
+    private hasPdfInsideIframe(iframe: HTMLIFrameElement): boolean {
+        const contentWindow: Window = iframe.contentWindow || ({length: 0} as Window);
+
+        if (typeof this.windowRef.chrome !== 'undefined') {
+            return contentWindow.length > 0;
+        } else if (isFirefox(this.windowRef.navigator.userAgent)) {
```
Comment (nsbarsukov) [inline projects/kit/components/pdf-viewer/pdf-viewer.component.ts:81]: use 
```ts
import {USER_AGENT} from '@ng-web-apis/common';
// ...
@Inject(USER_AGENT) private readonly userAgent: string,
```

File: projects/kit/components/pdf-viewer/pdf-viewer.component.ts [hunk 4]
```diff
@@ -1,33 +1,103 @@
...
+            return 'length' in contentWindow && 'then' in contentWindow;
+        } else if (isSafari(iframe)) {
+            return contentWindow.document.body.innerHTML.includes('embed');
```
Comment (nsbarsukov) [inline projects/kit/components/pdf-viewer/pdf-viewer.component.ts:84]: Could you please add some code-comments here ?

Honestly, i dont understand what happens here :(

**Review by nsbarsukov — CHANGES_REQUESTED** [review]
Your PR name: `feat(kit): detect pdf is not visible in browser`

Should we rename it to `fix(kit): ...` ?

File: projects/kit/components/pdf-viewer/pdf-viewer-options.ts [hunk 5]
```diff
@@ -2,6 +2,7 @@ import {PolymorpheusContent} from '@tinkoff/ng-polymorpheus';
 
 export interface TuiPdfViewerOptions<I = undefined> {
     readonly label: string;
+    readonly errorMessage?: string;
```
Comment (nsbarsukov) [inline projects/kit/components/pdf-viewer/pdf-viewer-options.ts:5]: Let's make it as just `error` (with type `PolymorpheusContent`).

Our users can decide to put some custom content.
For example, error-page component.

**Comment by vladimirpotekhin** [comment]
If user's browser setting "download pdf", he could download it and read before. Now he can't do it? (3th example)

**Comment by splincode** [comment]
https://github.com/Tinkoff/taiga-ui/pull/1920

**Pull request closed without merging**